### PR TITLE
Add HF adapter checkpoint tests and loss_compare checkpoint resume su…

### PR DESCRIPTION
**Pull request by mori360**
### Summary                                                         
                                                                                                             
  - Add unit tests for the HF save/load adapter integration in CheckpointManager, covering the  sd_adapter.to_hf/from_hf roundtrip that previously had no test coverage
  - Add --test-checkpoint-resume to loss_compare.py for verifying checkpoint save/load doesn't corrupt training state                                                                                             
  
  ### Test plan                                                                                                  
                                                                  
  - pytest tests/unit_tests/test_checkpoint.py -x — all tests pass (including 2 new HF adapter tests)        
  - pre-commit run --all-files — no new lint failures
  - loss_compare.py . . --test-checkpoint-resume --steps=20 --assert-equal — verify checkpoint resume produces identical loss                                                                                    
                                                                                                             
  ### Details                                                                                                    
                                                                  
  New checkpoint tests:                                                                                      
  - test_hf_save_adapter_receives_all_model_keys — verifies to_hf receives all model keys and dcp.save gets HF-keyed state dict                                                                                        
  - test_hf_load_roundtrip_to_hf_then_from_hf — verifies the full load sequence: to_hf (build container) → dcp.load → from_hf → load_state_dict, including call ordering and data flow through each step              
                                                                                                             
  loss_compare.py checkpoint resume:
  - --test-checkpoint-resume runs training in two phases: train to checkpoint interval and save, then resume to final step                                                                                             
  - --test-checkpoint-interval controls the save point (defaults to steps // 2)                              
  - Merged losses from both phases are compared against baseline to verify bit-exact resume

### CODE REVIEW

File: scripts/loss_compare.py [hunk 1]
```diff
@@ -992,6 +1005,24 @@ def parse_arguments() -> argparse.Namespace:
...
+    )
+    parser.add_argument(
+        "--test-checkpoint-interval",
+        type=int,
+        default=0,
+        help=(
+            "Checkpoint interval for --test-checkpoint-resume. "
+            "Defaults to steps // 2 if not specified."
+        ),
+    )
```
Comment (fegin) [inline scripts/loss_compare.py:1025]: Do we actually need this? This seems to be not necessary since checkpoint should always guarantee bit-wise parity. So we can just use the default interval?

**Review by fegin — COMMENTED** [review]
Do we plan to use `loss_compare.py` to do checkpoint integration test or no?

File: tests/unit_tests/test_checkpoint.py [hunk 2]
```diff
@@ -701,6 +701,121 @@ def fake_load(state_dict: dict, checkpoint_id=None):
...
+        to_hf_keys = set(mock_adapter.to_hf.call_args[0][0].keys())
+        self.assertEqual(to_hf_keys, {"weight", "bias"})
+
+        mock_adapter.from_hf.assert_called_once()
+        from_hf_keys = set(mock_adapter.from_hf.call_args[0][0].keys())
+        self.assertEqual(from_hf_keys, {"hf_weight", "hf_bias"})
+
+        mock_set_sd.assert_called_once()
+        loaded_sd = mock_set_sd.call_args[1]["model_state_dict"]
+        self.assertEqual(set(loaded_sd.keys()), {"weight", "bias"})
```
Comment (Andrei-Aksionov) [inline tests/unit_tests/test_checkpoint.py:816]: Not like this is super important, but it could be done simpler:
```python
from unittest.mock import ANY
...
mock_adapter.to_hf.assert_called_once_with({"weight": ANY, "bias": ANY})
mock_adapter.from_hf.assert_called_once_with({"hf_weight": ANY, "hf_bias": ANY})
mock_set_sd.assert_called_once_with(
    ANY, model_state_dict={"weight": ANY, "bias": ANY}, options=ANY
)

```

Btw, shouldn't this test also check that `get_hf_storage_reader` was called?

File: tests/unit_tests/test_checkpoint.py [hunk 3]
```diff
@@ -701,6 +701,121 @@ def fake_load(state_dict: dict, checkpoint_id=None):
...
+        manager.save(curr_step=10, last_step=True)
+
+        mock_adapter.to_hf.assert_called_once()
+        self.assertIn("weight", received_keys)
+        self.assertIn("bias", received_keys)
+
+        mock_save.assert_called_once()
+        saved_sd = mock_save.call_args[0][0]
+        self.assertIn("hf_weight", saved_sd)
+        self.assertIn("hf_bias", saved_sd)
```
Comment (Andrei-Aksionov) [inline tests/unit_tests/test_checkpoint.py:749]: I guess it's better to do a strict comparison instead of `assertIn`.

```python
mock_adapter.to_hf.assert_called_once_with({"weight": ANY, "bias": ANY})

mock_save.assert_called_once()
self.assertEqual(
    mock_save.call_args.args[0], {"hf_weight": ANY, "hf_bias": ANY}
)
```

(With that `recieved_keys` is no longer needed.)

File: scripts/loss_compare.py (hunk 1 above)
Comment (Andrei-Aksionov) [reply scripts/loss_compare.py:1025]: Hello @fegin 
Could you please clarify what did you mean? 😊 

> This seems to be not necessary since checkpoint should always guarantee bit-wise parity.

I haven't noticed anywhere in the code that verifies bit-wise parity.
I assume that you've meant that PyTorch DCP gives that guarantee, but since `CheckpointManager` class is basically a wrapper around it, it might be a good idea to check the correctness via this integration test.

That should also make it easier to give answer to questions like https://github.com/pytorch/torchtitan/pull/2839#pullrequestreview-4059108990 whenever there are significant changes to the checkpointing logic.
